[Converter] Added memory base buffer; Code clean-up

Zyiqin-Miranda opened the pull request:
## Summary

This PR adjust the memory resource estimation for converter() remote tasks based off experiments. Also contains code clean-up regarding determining which snapshot type to commit. 

All existing tests passing.
## Testing

Describe how the changes have been tested, including both automated and manual testing strategies.
If this is a bugfix, explain how the fix has been tested to ensure the bug is resolved without introducing new issues.

## Regression Risk

If this is a bugfix, assess the risk of regression caused by this fix and steps taken to mitigate it.

## Checklist

- [ ] Unit tests covering the changes have been added
  - [ ] If this is a bugfix, regression tests have been added

- [ ] E2E testing has been performed

## Additional Notes

Any additional information or context relevant to this PR.

pdames commented on deltacat/compute/converter/utils/converter_session_utils.py:
```diff
@@ -121,3 +123,53 @@ def sort_data_files_maintaining_order(data_files: DataFileList) -> DataFileList:
+    DELETE = "delete"
+
+
+def _get_snapshot_description(
```
Since this isn't describing the snapshot but instead the action that we'll take based on the snapshot, maybe something like:
```suggestion
def _get_snapshot_action_description(
```

pdames commented on deltacat/compute/converter/utils/convert_task_options.py:
```diff
@@ -3,13 +3,14 @@
-# TODO: Add audit info to check this number in practice
 # Worst case 2 as no duplicates exists across all pk
 PYARROW_AGGREGATE_MEMORY_MULTIPLIER = 2
+# Observed base memory usage at the beginning of each process
```
```suggestion
# Observed base memory usage at the beginning of each worker process
```

pdames commented on deltacat/compute/converter/utils/convert_task_options.py:
```diff
@@ -3,13 +3,14 @@
 # Worst case 2 as no duplicates exists across all pk
 PYARROW_AGGREGATE_MEMORY_MULTIPLIER = 2
+# Observed base memory usage at the beginning of each process
+BASE_MEMORY_BUFFER = 0.3 * 1024 * 1024 * 1024
```
Should we also document that we need at least (300MB + some buffer) of free memory to run the converter? Maybe just the docs for this function should state that the converter requires a node with at least 512MB of free memory?

Zyiqin-Miranda replied:
Thanks @pdames, added as `**Memory Requirements:**` doc string in `converter_session()` function. Haven't finalize the naming of al params like maybe `primary key` to `merge key` yet, so doc string is still using the current version of converter. 
Additionally, tested locally that even with just test setup, the process memory is using around 300MB to begin with. Just FYI.

pdames commented on deltacat/utils/resources.py:
```diff
@@ -169,6 +170,12 @@ def get_current_process_peak_memory_usage_in_bytes():
         return psutil.Process().memory_info().peak_wset
 
 
+def get_current_process_current_memory_usage_in_bytes():
+    process = psutil.Process(os.getpid())
+    mem_info = process.memory_info()
+    return mem_info
+
+
```
Since this isn't currently used can we remove it? Or if we're going to use it for a profiler in a follow-up PR we can keep it for now.

pdames reviewed (approved):
Looks good overall! Just a few minor comments.

Zyiqin-Miranda commented:
Thanks @pdames, addressed comments and will follow up soon with two PRs: 
1. For make all private functions start with underscore with additional code clean-ups
2. Try-catch OOM failures and retry with additional resources, probably will test with removing custom resources like "coverter_task": 1 to use just CPU resources to make cluster configurations more generic.

Zyiqin-Miranda commented:
Also, removed the `comment-on-alert: true` when performance testing triggered alert since write permission in general shouldn't be granted with fork, from [github actions security doc](https://docs.github.com/en/actions/how-tos/security-for-github-actions/security-guides/automatic-token-authentication): 
>`You can use the permissions key to add and remove read permissions for forked repositories, but typically you can't grant write access`
So probably should setup another way of alerting on action workflow benchmark testing regression like email or slack notification like other repos do. But removing this comment option in CI for now to unblock contribution merging.
